File: membrain-py/membrain/cbr/retriever.py
"""Case retrieval strategies for case-based reasoning.

Provides an abstract CaseRetriever base class and a NonParametricRetriever
that uses embedding similarity via MembrainClient.search_cases().
"""

from __future__ import annotations

from abc import ABC, abstractmethod

from ..client import MembrainClient
from ..types import CaseSearchResults


class CaseRetriever(ABC):
    """Abstract base class for case retrieval strategies."""

    @abstractmethod
    async def retrieve(
        self,
        query: str,
        limit: int = 5,
        positive_reward_threshold: float = 0.5,
    ) -> CaseSearchResults:
        """Retrieve similar cases for a given query.

        Args:
            query: The problem description to find similar cases for.
            limit: Maximum number of cases to return.
            positive_reward_threshold: Reward threshold separating
                positive from negative cases.

        Returns:
            CaseSearchResults with positive and negative case lists.
        """


class NonParametricRetriever(CaseRetriever):
    """Embedding-similarity retriever using MembrainClient.search_cases().

    No training required -- works immediately with stored cases by using
    cosine similarity on embeddings.
    """

    def __init__(self, client: MembrainClient) -> None:
        self._client = client

    async def retrieve(
        self,
        query: str,
        limit: int = 5,
        positive_reward_threshold: float = 0.5,
    ) -> CaseSearchResults:
        """Retrieve cases ranked by embedding similarity."""
        return await self._client.search_cases(
            query,
            limit=limit,
            positive_reward_threshold=positive_reward_threshold,
        )

File: membrain-node/src/cbr/retriever.ts
/**
 * Case retrieval strategies for case-based reasoning.
 *
 * Exposes an abstract {@link CaseRetriever} base class and a
 * {@link NonParametricRetriever} that uses embedding similarity via
 * {@link MembrainClient.searchCases}.
 */

import type { MembrainClient } from "../client";
import type { CaseSearchResults } from "../types";

export abstract class CaseRetriever {
  abstract retrieve(
    query: string,
    limit?: number,
    positiveRewardThreshold?: number
  ): Promise<CaseSearchResults> | CaseSearchResults;
}

export class NonParametricRetriever extends CaseRetriever {
  private readonly client: MembrainClient;

  constructor(client: MembrainClient) {
    super();
    this.client = client;
  }

  retrieve(
    query: string,
    limit: number = 5,
    positiveRewardThreshold: number = 0.5
  ): CaseSearchResults {
    return this.client.searchCases(query, limit, positiveRewardThreshold);
  }
}
